gh-95451: Update docs for wasm32-emscripten and -wasi platforms (#95452)
python/cpython: Doc/library/intro.rst

  libc version, then both conditions must hold. For example a feature with note
  *Availability: Linux >= 3.17 with glibc >= 2.27* requires both Linux 3.17 or
  newer and glibc 2.27 or newer.

.. _wasm-availability:

Notes on WebAssembly platforms
==============================

The WebAssembly platforms ``wasm32-emscripten`` and ``wasm32-wasi`` provide a

> subset of POSIX APIs. WASM runtimes and browser are sandboxed and have

Reviewer: ```suggestion
subset of POSIX APIs. WASM runtimes and browsers are sandboxed and have
```

> limited access to host and rest of the world. Any standard library module,

Reviewer 2: ```suggestion
limited access to the host and external resources. Any standard library module
```

> that uses processes, threading, networking (:mod:`socket`), signals, or

Reviewer 3: Is there a reason why only the `socket` module is specified and not other modules?

Author: Which other modules do you expect? All networking modules are based on the socket module.

Reviewer 3: I just found it a bit odd that there are no mentions about specific modules for processes (e.g. `subprocess`, `multiprocessing`) or threading, and one single example for networking.

Author: Makes sense. I have removed the reference to the socket module from the sentence. Initially I wanted to include examples for each missing feature, but then it turned out to make the sense much harder to read.

other forms of inter-process communication (IPC), is either not available
or may not work as on other Unix-like systems. File I/O, file system, and Unix

> permission-related functions are restricted, too. Emscripten does not permit

Reviewer 3: Possibly part of a separate PR, but a few glossary entries about WebAssembly, Emscripten, and related terms would be useful IMHO.

> blocking operations like blocking network I/O and :func:`~time.sleep`.

Reviewer 3: I would rephrase this to avoid the double "blocking".

The properties and behavior of Python on WebAssembly platforms depend on the
`Emscripten`_ or `WASI`_-SDK version, WASM runtimes (browser, NodeJS, `wasmtime`_),
and Python build time flags. WebAssembly, Emscripten, and WASI are evolving.

> Some features like networking may be supported in the future.

Reviewer: Given that Emscripten and WASI runtimes are moving targets, there could at least be unofficial builds of Python build against future versions of emscripten where some of the features marked as unavailable become available.  Do we want to somehow flag the difference between "explicitly disabled in Python" vs. "not known to work in current environments"?  *Maybe* we want to add to this paragraph:

"Python features documented as unsupported may in fact work depending on your environment."

But *maybe* that's too confusing for most users to understand, and something we can address in the future as these things mature.

Author: Very few features are explicitly disabled. The ``config.site`` files disable a few syscalls that fail tests or make no sense. For example I disabled the effective user/group id function family. Emscripten returns 0, which some code interprets as `CAP_SYS_ADMIN` privileges. The `config.site` files in `Tool/wasm` directory have comments for special cases.

The target audience for the section is "curious users without previous WASM experience". I want to give them *some* ideas what does not work without going into too many technical details.

For Python in the browser, users should consider `Pyodide`_ or `PyScript`_.
PyScript is built on top of Pyodide, which itself is built on top of
CPython and Emscripten. Pyodide provides access to browsers' JavaScript and
DOM APIs as well as limited networking capabilities with JavaScript's

> ``XMLHttpRequest`` and ``Fetch`` API.

Reviewer: Looking forward to the day when that's just a "browser support module" installable in upstream CPython...

Author: I'm lobbying for a high level HTTP API in Python core.

> ``XMLHttpRequest`` and ``Fetch`` API.

Reviewer 3: ```suggestion
``XMLHttpRequest`` and ``Fetch`` APIs.
```

.. TODO: update with information from Tools/wasm/README.md

* Process-related APIs are not available or always fail with an error. That
  includes APIs that spawn new processes (:func:`~os.fork`,
  :func:`~os.execve`), wait for processes (:func:`~os.waitpid`), send signals
  (:func:`~os.kill`), or otherwise interact with processes. The
  :mod:`subprocess` is importable but does not work.

> * The :mod:`socket` module is available, but is dysfunctional.

Reviewer 3: "Dysfunctional" is not very informative -- could you elaborate a bit?

> * Some functions are stubs that either don't do anything and always return some

Reviewer: ```suggestion
* Some functions are stubs that either don't do anything or always return some
```

> * Some functions are stubs that either don't do anything and always return some
>   hardcoded value.

Reviewer 3: ```suggestion
* Some functions are stubs that either don't do anything and always return
  hardcoded values.
```

  libc version, then both conditions must hold. For example a feature with note
  *Availability: Linux >= 3.17 with glibc >= 2.27* requires both Linux 3.17 or
  newer and glibc 2.27 or newer.

.. _wasm-availability:

> Notes on WebAssembly platforms
> ==============================

Reviewer 4: ```suggestion
WebAssembly platforms
---------------------
```

I'd suggest making this a subsection of the top-level "Notes on availability" section, since this is a subtopic of it (availability -> WASM availability), and it also makes it a part of the section users reach when they click the link in the ``availability`` directive, which is now heavily being used for WASM (since it links to the "Availability" section) rather than a separate parallel section. It also reduces redundancy in the title.

The `WebAssembly`_ platforms ``wasm32-emscripten`` (`Emscripten`_) and

> ``wasm32-wasi`` `WASI`_ provide a subset of POSIX APIs. WebAssembly runtimes

Reviewer 4: ```suggestion
``wasm32-wasi`` (`WASI`_) provide a subset of POSIX APIs. WebAssembly runtimes
```

Be consistent with previous

and browsers are sandboxed and have limited access to the host and external
resources. Any Python standard library module that uses processes, threading,
networking, signals, or other forms of inter-process communication (IPC), is
either not available or may not work as on other Unix-like systems. File I/O,
file system, and Unix permission-related functions are restricted, too.
Emscripten does not permit blocking I/O or other blocking operations like
:func:`~time.sleep`.

The properties and behavior of Python on WebAssembly platforms depend on the
`Emscripten`_-SDK or `WASI`_-SDK version, WASM runtimes (browser, NodeJS,
`wasmtime`_), and Python build time flags. WebAssembly, Emscripten, and WASI

> are evolving and moving standards. Some features like networking may be

Reviewer 4: ```suggestion
are evolving standards; some features like networking may be
```

* Avoid redundant/verbose word ("moving" with "evolving")
* Join with semicolon to reduce choppiness and improve flow

supported in the future.

For Python in the browser, users should consider `Pyodide`_ or `PyScript`_.
PyScript is built on top of Pyodide, which itself is built on top of
CPython and Emscripten. Pyodide provides access to browsers' JavaScript and
DOM APIs as well as limited networking capabilities with JavaScript's
``XMLHttpRequest`` and ``Fetch`` APIs.

* Process-related APIs are not available or always fail with an error. That
  includes APIs that spawn new processes (:func:`~os.fork`,
  :func:`~os.execve`), wait for processes (:func:`~os.waitpid`), send signals
  (:func:`~os.kill`), or otherwise interact with processes. The
  :mod:`subprocess` is importable but does not work.

> * The :mod:`socket` module is available, but is dysfunctional. On Emscripten,

Reviewer 4: ```suggestion
* The :mod:`socket` module is available,
but is limited and behaves differently from other platforms.
On Emscripten,
```

Be more specific and avoid charged and potentially confusing words like "dysfunctional"

  sockets are always non-blocking and require additional JavaScript code

>   and helpers on the server to proxy TCP through WebSockets, see

Reviewer 4: ```suggestion
  and helpers on the server to proxy TCP through WebSockets; see
```

Fix comma splice

>   `Emscripten Networking <https://emscripten.org/docs/porting/networking.html>`_

Reviewer 4: Minor style nitpick, but is there a reason this link was added inline, but all the others were not?

  for more information. WASI snapshot preview 1 only permits sockets from an
  existing file descriptor.

* Some functions are stubs that either don't do anything and always return

>   hardcoded values.

Reviewer 4: Did you mean to refer to, e.g., the ``os`` module here? Another specific module? Or "other modules" in general? Whatever is the case, it should be explicitly specified, as I was rather confused what this was supposed to say. Furthermore, if "other modules" in general (as currently implied), it should be the last bullet, since it is a catchall that covers everything. Also, did did you mean "either ... or", or just "and" (no "either")? Right now it reads "either ... and" which doesn't really make grammatical sense.

* Functions related to file descriptors, file permissions, file ownership, and

>   links are limited and don't support some operations. For example WASI does

Reviewer 4: ```suggestion
  links are limited and don't support some operations. For example, WASI does
```

Add missing comma
